File: algorithms.js
/**
 * Graph Algorithms Implementation
 * Including: BFS, DFS, Dijkstra's Algorithm, and A* Algorithm
 */

class GraphAlgorithms {
    /**
     * Breadth-First Search (BFS)
     * Finds shortest path in terms of number of edges (unweighted)
     */
    static bfs(graph, startId, endId) {
        if (!graph.nodes.has(startId) || !graph.nodes.has(endId)) {
            return null;
        }

        const queue = [{nodeId: startId, path: [startId]}];
        const visited = new Set([startId]);

        while (queue.length > 0) {
            const {nodeId, path} = queue.shift();

            if (nodeId === endId) {
                return {
                    path: path.map(id => graph.getNode(id)),
                    distance: path.length - 1,
                    algorithm: 'BFS'
                };
            }

            const neighbors = graph.getNeighbors(nodeId);
            for (const {node} of neighbors) {
                if (!visited.has(node.id)) {
                    visited.add(node.id);
                    queue.push({
                        nodeId: node.id,
                        path: [...path, node.id]
                    });
                }
            }
        }

        return null; // No path found
    }

    /**
     * Depth-First Search (DFS)
     * Finds a path (not necessarily shortest)
     */
    static dfs(graph, startId, endId) {
        if (!graph.nodes.has(startId) || !graph.nodes.has(endId)) {
            return null;
        }

        const visited = new Set();
        const path = [];

        const dfsRecursive = (currentId) => {
            visited.add(currentId);
            path.push(currentId);

            if (currentId === endId) {
                return true;
            }

            const neighbors = graph.getNeighbors(currentId);
            for (const {node} of neighbors) {
                if (!visited.has(node.id)) {
                    if (dfsRecursive(node.id)) {
                        return true;
                    }
                }
            }

            path.pop();
            return false;
        };

        if (dfsRecursive(startId)) {
            return {
                path: path.map(id => graph.getNode(id)),
                distance: path.length - 1,
                algorithm: 'DFS'
            };
        }

        return null; // No path found
    }

    /**
     * Dijkstra's Algorithm
     * Finds shortest path in weighted graphs
     */
    static dijkstra(graph, startId, endId, weightProperty = 'weight') {
        if (!graph.nodes.has(startId) || !graph.nodes.has(endId)) {
            return null;
        }

        const distances = new Map();
        const previous = new Map();
        const unvisited = new Set();

        // Initialize distances
        for (const nodeId of graph.nodes.keys()) {
            distances.set(nodeId, Infinity);
            previous.set(nodeId, null);
            unvisited.add(nodeId);
        }
        distances.set(startId, 0);

        while (unvisited.size > 0) {
            // Find unvisited node with minimum distance
            let currentId = null;
            let minDistance = Infinity;
            for (const nodeId of unvisited) {
                if (distances.get(nodeId) < minDistance) {
                    minDistance = distances.get(nodeId);
                    currentId = nodeId;
                }
            }

            if (currentId === null || minDistance === Infinity) {
                break; // No path exists
            }

            unvisited.delete(currentId);

            // If we reached the destination
            if (currentId === endId) {
                break;
            }

            // Update distances to neighbors
            const neighbors = graph.getNeighbors(currentId);
            for (const {node, edge} of neighbors) {
                if (unvisited.has(node.id)) {
                    const weight = this.getWeight(edge, weightProperty);
                    const altDistance = distances.get(currentId) + weight;

                    if (altDistance < distances.get(node.id)) {
                        distances.set(node.id, altDistance);
                        previous.set(node.id, currentId);
                    }
                }
            }
        }

        // Reconstruct path
        if (distances.get(endId) === Infinity) {
            return null; // No path found
        }

        const path = [];
        let current = endId;
        while (current !== null) {
            path.unshift(current);
            current = previous.get(current);
        }

        return {
            path: path.map(id => graph.getNode(id)),
            distance: distances.get(endId),
            algorithm: 'Dijkstra',
            weightProperty: weightProperty
        };
    }

    /**
     * A* Algorithm
     * Finds shortest path using heuristic (requires coordinates)
     */
    static aStar(graph, startId, endId, weightProperty = 'weight') {
        if (!graph.nodes.has(startId) || !graph.nodes.has(endId)) {
            return null;
        }

        const startNode = graph.getNode(startId);
        const endNode = graph.getNode(endId);

        // Check if nodes have coordinates for heuristic
        if (!startNode.data.x || !endNode.data.x) {
            console.warn('A* requires node coordinates, falling back to Dijkstra');
            return this.dijkstra(graph, startId, endId, weightProperty);
        }

        const openSet = new Set([startId]);
        const closedSet = new Set();
        const gScore = new Map(); // Cost from start to node
        const fScore = new Map(); // gScore + heuristic
        const previous = new Map();

        // Initialize scores
        for (const nodeId of graph.nodes.keys()) {
            gScore.set(nodeId, Infinity);
            fScore.set(nodeId, Infinity);
            previous.set(nodeId, null);
        }
        gScore.set(startId, 0);
        fScore.set(startId, this.heuristic(startNode, endNode));

        while (openSet.size > 0) {
            // Find node in openSet with lowest fScore
            let currentId = null;
            let minFScore = Infinity;
            for (const nodeId of openSet) {
                if (fScore.get(nodeId) < minFScore) {
                    minFScore = fScore.get(nodeId);
                    currentId = nodeId;
                }
            }

            if (currentId === null) {
                break;
            }

            // If we reached the destination
            if (currentId === endId) {
                const path = [];
                let current = endId;
                while (current !== null) {
                    path.unshift(current);
                    current = previous.get(current);
                }

                return {
                    path: path.map(id => graph.getNode(id)),
                    distance: gScore.get(endId),
                    algorithm: 'A*',
                    weightProperty: weightProperty
                };
            }

            openSet.delete(currentId);
            closedSet.add(currentId);

            // Check neighbors
            const neighbors = graph.getNeighbors(currentId);
            for (const {node, edge} of neighbors) {
                if (closedSet.has(node.id)) {
                    continue;
                }

                const weight = this.getWeight(edge, weightProperty);
                const tentativeGScore = gScore.get(currentId) + weight;

                if (!openSet.has(node.id)) {
                    openSet.add(node.id);
                } else if (tentativeGScore >= gScore.get(node.id)) {
                    continue;
                }

                // This path is the best so far
                previous.set(node.id, currentId);
                gScore.set(node.id, tentativeGScore);
                fScore.set(node.id, tentativeGScore + this.heuristic(node, endNode));
            }
        }

        return null; // No path found
    }

    /**
     * Heuristic function for A* (Euclidean distance)
     */
    static heuristic(nodeA, nodeB) {
        const dx = nodeA.data.x - nodeB.data.x;
        const dy = nodeA.data.y - nodeB.data.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    /**
     * Get weight from edge based on property
     */
    static getWeight(edge, weightProperty) {
        if (weightProperty === 'weight') {
            return edge.weight;
        }
        return edge.data[weightProperty] || edge.weight;
    }

    /**
     * Find all paths between two nodes (DFS-based)
     */
    static findAllPaths(graph, startId, endId, maxPaths = 10) {
        if (!graph.nodes.has(startId) || !graph.nodes.has(endId)) {
            return [];
        }

        const allPaths = [];
        const visited = new Set();

        const dfsAllPaths = (currentId, path) => {
            if (allPaths.length >= maxPaths) {
                return;
            }

            if (currentId === endId) {
                allPaths.push([...path]);
                return;
            }

            visited.add(currentId);

            const neighbors = graph.getNeighbors(currentId);
            for (const {node} of neighbors) {
                if (!visited.has(node.id)) {
                    path.push(node.id);
                    dfsAllPaths(node.id, path);
                    path.pop();
                }
            }

            visited.delete(currentId);
        };

        dfsAllPaths(startId, [startId]);

        return allPaths.map(path => ({
            path: path.map(id => graph.getNode(id)),
            distance: path.length - 1
        }));
    }

    /**
     * Calculate total cost of a path based on weight property
     */
    static calculatePathCost(graph, path, weightProperty = 'weight') {
        let totalCost = 0;
        for (let i = 0; i < path.length - 1; i++) {
            const currentId = path[i].id;
            const nextId = path[i + 1].id;
            
            const neighbors = graph.getNeighbors(currentId);
            const neighbor = neighbors.find(({node}) => node.id === nextId);
            
            if (neighbor) {
                totalCost += this.getWeight(neighbor.edge, weightProperty);
            }
        }
        return totalCost;
    }
}

File: algorithms.c
/**
 * Pathfinding Algorithms Implementation
 * Travel Route Planner - C Implementation
 * Implements: Dijkstra, A*, BFS, DFS
 */

#include "graph.h"

// Priority queue node for Dijkstra and A*
typedef struct PQNode {
    int cityIndex;
    int priority;
    struct PQNode* next;
} PQNode;

// Priority queue functions
PQNode* createPQNode(int cityIndex, int priority) {
    PQNode* node = (PQNode*)malloc(sizeof(PQNode));
    node->cityIndex = cityIndex;
    node->priority = priority;
    node->next = NULL;
    return node;
}

void enqueuePQ(PQNode** head, int cityIndex, int priority) {
    PQNode* newNode = createPQNode(cityIndex, priority);
    
    if (*head == NULL || priority < (*head)->priority) {
        newNode->next = *head;
        *head = newNode;
    } else {
        PQNode* current = *head;
        while (current->next && current->next->priority <= priority) {
            current = current->next;
        }
        newNode->next = current->next;
        current->next = newNode;
    }
}

int dequeuePQ(PQNode** head) {
    if (*head == NULL) return -1;
    
    PQNode* temp = *head;
    int cityIndex = temp->cityIndex;
    *head = (*head)->next;
    free(temp);
    return cityIndex;
}

bool isPQEmpty(PQNode* head) {
    return head == NULL;
}

/**
 * Dijkstra's Algorithm
 * Finds shortest path in weighted graph
 */
PathResult* dijkstra(Graph* graph, const char* startId, const char* endId, const char* weightType) {
    int start = findCityIndex(graph, startId);
    int end = findCityIndex(graph, endId);
    
    if (start == -1 || end == -1) {
        return NULL;
    }
    
    int n = graph->numCities;
    int* dist = (int*)malloc(n * sizeof(int));
    int* parent = (int*)malloc(n * sizeof(int));
    bool* visited = (bool*)malloc(n * sizeof(bool));
    
    // Initialize
    for (int i = 0; i < n; i++) {
        dist[i] = INFINITY_DIST;
        parent[i] = -1;
        visited[i] = false;
    }
    
    dist[start] = 0;
    PQNode* pq = NULL;
    enqueuePQ(&pq, start, 0);
    
    // Main algorithm
    while (!isPQEmpty(pq)) {
        int current = dequeuePQ(&pq);
        
        if (visited[current]) continue;
        visited[current] = true;
        
        if (current == end) break;
        
        // Explore neighbors
        AdjNode* neighbor = graph->adjList[current];
        while (neighbor) {
            int next = neighbor->cityIndex;
            int weight = getWeight(neighbor, weightType);
            int newDist = dist[current] + weight;
            
            if (newDist < dist[next]) {
                dist[next] = newDist;
                parent[next] = current;
                enqueuePQ(&pq, next, newDist);
            }
            
            neighbor = neighbor->next;
        }
    }
    
    // Check if path exists
    if (dist[end] == INFINITY_DIST) {
        free(dist);
        free(parent);
        free(visited);
        return NULL;
    }
    
    // Reconstruct path
    PathResult* result = (PathResult*)malloc(sizeof(PathResult));
    strcpy(result->algorithm, "Dijkstra's Algorithm");
    result->pathLength = 0;
    
    int temp[MAX_CITIES];
    int tempLen = 0;
    int current = end;
    
    while (current != -1) {
        temp[tempLen++] = current;
        current = parent[current];
    }
    
    // Reverse path
    for (int i = 0; i < tempLen; i++) {
        result->path[i] = temp[tempLen - 1 - i];
    }
    result->pathLength = tempLen;
    
    // Calculate metrics
    result->totalDistance = 0;
    result->totalTime = 0;
    result->totalCost = 0;
    
    for (int i = 0; i < result->pathLength - 1; i++) {
        int from = result->path[i];
        int to = result->path[i + 1];
        
        AdjNode* node = graph->adjList[from];
        while (node && node->cityIndex != to) {
            node = node->next;
        }
        
        if (node) {
            result->totalDistance += node->distance;
            result->totalTime += node->time;
            result->totalCost += node->cost;
        }
    }
    
    free(dist);
    free(parent);
    free(visited);
    
    return result;
}

/**
 * A* Algorithm
 * Enhanced Dijkstra with heuristic
 */
PathResult* aStar(Graph* graph, const char* startId, const char* endId, const char* weightType) {
    int start = findCityIndex(graph, startId);
    int end = findCityIndex(graph, endId);
    
    if (start == -1 || end == -1) {
        return NULL;
    }
    
    int n = graph->numCities;
    int* gScore = (int*)malloc(n * sizeof(int));
    int* fScore = (int*)malloc(n * sizeof(int));
    int* parent = (int*)malloc(n * sizeof(int));
    bool* visited = (bool*)malloc(n * sizeof(bool));
    
    // Initialize
    for (int i = 0; i < n; i++) {
        gScore[i] = INFINITY_DIST;
        fScore[i] = INFINITY_DIST;
        parent[i] = -1;
        visited[i] = false;
    }
    
    gScore[start] = 0;
    fScore[start] = (int)calculateHeuristic(graph, start, end);
    
    PQNode* pq = NULL;
    enqueuePQ(&pq, start, fScore[start]);
    
    // Main algorithm
    while (!isPQEmpty(pq)) {
        int current = dequeuePQ(&pq);
        
        if (visited[current]) continue;
        visited[current] = true;
        
        if (current == end) break;
        
        // Explore neighbors
        AdjNode* neighbor = graph->adjList[current];
        while (neighbor) {
            int next = neighbor->cityIndex;
            int weight = getWeight(neighbor, weightType);
            int tentativeG = gScore[current] + weight;
            
            if (tentativeG < gScore[next]) {
                parent[next] = current;
                gScore[next] = tentativeG;
                fScore[next] = tentativeG + (int)calculateHeuristic(graph, next, end);
                enqueuePQ(&pq, next, fScore[next]);
            }
            
            neighbor = neighbor->next;
        }
    }
    
    // Check if path exists
    if (gScore[end] == INFINITY_DIST) {
        free(gScore);
        free(fScore);
        free(parent);
        free(visited);
        return NULL;
    }
    
    // Reconstruct path (same as Dijkstra)
    PathResult* result = (PathResult*)malloc(sizeof(PathResult));
    strcpy(result->algorithm, "A* Algorithm");
    result->pathLength = 0;
    
    int temp[MAX_CITIES];
    int tempLen = 0;
    int current = end;
    
    while (current != -1) {
        temp[tempLen++] = current;
        current = parent[current];
    }
    
    for (int i = 0; i < tempLen; i++) {
        result->path[i] = temp[tempLen - 1 - i];
    }
    result->pathLength = tempLen;
    
    // Calculate metrics
    result->totalDistance = 0;
    result->totalTime = 0;
    result->totalCost = 0;
    
    for (int i = 0; i < result->pathLength - 1; i++) {
        int from = result->path[i];
        int to = result->path[i + 1];
        
        AdjNode* node = graph->adjList[from];
        while (node && node->cityIndex != to) {
            node = node->next;
        }
        
        if (node) {
            result->totalDistance += node->distance;
            result->totalTime += node->time;
            result->totalCost += node->cost;
        }
    }
    
    free(gScore);
    free(fScore);
    free(parent);
    free(visited);
    
    return result;
}

/**
 * Breadth-First Search (BFS)
 * Finds path with minimum number of stops
 */
PathResult* bfs(Graph* graph, const char* startId, const char* endId) {
    int start = findCityIndex(graph, startId);
    int end = findCityIndex(graph, endId);
    
    if (start == -1 || end == -1) {
        return NULL;
    }
    
    int n = graph->numCities;
    bool* visited = (bool*)malloc(n * sizeof(bool));
    int* parent = (int*)malloc(n * sizeof(int));
    int* queue = (int*)malloc(n * sizeof(int));
    
    for (int i = 0; i < n; i++) {
        visited[i] = false;
        parent[i] = -1;
    }
    
    int front = 0, rear = 0;
    queue[rear++] = start;
    visited[start] = true;
    
    bool found = false;
    
    while (front < rear && !found) {
        int current = queue[front++];
        
        if (current == end) {
            found = true;
            break;
        }
        
        AdjNode* neighbor = graph->adjList[current];
        while (neighbor) {
            int next = neighbor->cityIndex;
            if (!visited[next]) {
                visited[next] = true;
                parent[next] = current;
                queue[rear++] = next;
            }
            neighbor = neighbor->next;
        }
    }
    
    if (!found) {
        free(visited);
        free(parent);
        free(queue);
        return NULL;
    }
    
    // Reconstruct path
    PathResult* result = (PathResult*)malloc(sizeof(PathResult));
    strcpy(result->algorithm, "Breadth-First Search (BFS)");
    
    int temp[MAX_CITIES];
    int tempLen = 0;
    int current = end;
    
    while (current != -1) {
        temp[tempLen++] = current;
        current = parent[current];
    }
    
    for (int i = 0; i < tempLen; i++) {
        result->path[i] = temp[tempLen - 1 - i];
    }
    result->pathLength = tempLen;
    
    // Calculate metrics
    result->totalDistance = 0;
    result->totalTime = 0;
    result->totalCost = 0;
    
    for (int i = 0; i < result->pathLength - 1; i++) {
        int from = result->path[i];
        int to = result->path[i + 1];
        
        AdjNode* node = graph->adjList[from];
        while (node && node->cityIndex != to) {
            node = node->next;
        }
        
        if (node) {
            result->totalDistance += node->distance;
            result->totalTime += node->time;
            result->totalCost += node->cost;
        }
    }
    
    free(visited);
    free(parent);
    free(queue);
    
    return result;
}

/**
 * Depth-First Search (DFS) Helper
 */
bool dfsHelper(Graph* graph, int current, int end, bool* visited, int* parent) {
    visited[current] = true;
    
    if (current == end) {
        return true;
    }
    
    AdjNode* neighbor = graph->adjList[current];
    while (neighbor) {
        int next = neighbor->cityIndex;
        if (!visited[next]) {
            parent[next] = current;
            if (dfsHelper(graph, next, end, visited, parent)) {
                return true;
            }
        }
        neighbor = neighbor->next;
    }
    
    return false;
}

/**
 * Depth-First Search (DFS)
 * Finds any path (may not be optimal)
 */
PathResult* dfs(Graph* graph, const char* startId, const char* endId) {
    int start = findCityIndex(graph, startId);
    int end = findCityIndex(graph, endId);
    
    if (start == -1 || end == -1) {
        return NULL;
    }
    
    int n = graph->numCities;
    bool* visited = (bool*)malloc(n * sizeof(bool));
    int* parent = (int*)malloc(n * sizeof(int));
    
    for (int i = 0; i < n; i++) {
        visited[i] = false;
        parent[i] = -1;
    }
    
    bool found = dfsHelper(graph, start, end, visited, parent);
    
    if (!found) {
        free(visited);
        free(parent);
        return NULL;
    }
    
    // Reconstruct path
    PathResult* result = (PathResult*)malloc(sizeof(PathResult));
    strcpy(result->algorithm, "Depth-First Search (DFS)");
    
    int temp[MAX_CITIES];
    int tempLen = 0;
    int current = end;
    
    while (current != -1) {
        temp[tempLen++] = current;
        current = parent[current];
    }
    
    for (int i = 0; i < tempLen; i++) {
        result->path[i] = temp[tempLen - 1 - i];
    }
    result->pathLength = tempLen;
    
    // Calculate metrics
    result->totalDistance = 0;
    result->totalTime = 0;
    result->totalCost = 0;
    
    for (int i = 0; i < result->pathLength - 1; i++) {
        int from = result->path[i];
        int to = result->path[i + 1];
        
        AdjNode* node = graph->adjList[from];
        while (node && node->cityIndex != to) {
            node = node->next;
        }
        
        if (node) {
            result->totalDistance += node->distance;
            result->totalTime += node->time;
            result->totalCost += node->cost;
        }
    }
    
    free(visited);
    free(parent);
    
    return result;
}
